### PR TITLE
add a test case with gated pod-affinity pods to scheduler_perf

### PR DESCRIPTION
**Pull request by sanposhiho**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

add a test case with gated pods to scheduler_perf to ensure https://github.com/kubernetes/kubernetes/issues/124384 isn't happening.

#### Which issue(s) this PR fixes:

Part of #124384

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by gabesaba** [comment]
I don't think this test will reproduce the performance issue, as it happens when there are gated pods in conjunction with many cluster events. Please see https://github.com/kubernetes/kubernetes/pull/125504, which reproduces that case.

**Comment by sanposhiho** [comment]
It does for sure, see https://github.com/kubernetes/kubernetes/pull/125527#discussion_r1642925743.

> it happens when there are gated pods in conjunction with many cluster events.

Pods getting scheduled are cluster events.

**Comment by sanposhiho** [comment]
/hold

Actually, let me double-confirm. (I ran the test twice for master and in #125527, but ran the test for a revert case only once; I know scheduler_perf sometimes show much slower results somehow.)

**Comment by sanposhiho** [comment]
/unhold

My previous version wasn't great after all and then I had to improve it to make it stable to reproduce the degradation.

---

Also, while I was at it, I found another (minor?) scenario that many gated Pods could slow down the throughput.. This one could be reproduced with `SchedulingGatedPodsWithPodAffinityImpactForThroughput`.

The degradation could happen when all these conditions are met:
- There are many gated Pods with PodAffinity.
- There are many Pods getting scheduled, that is, the scheduling queue keeps receiving AssignedPodAdd cluster events.

After some debugging, I found `getUnschedulablePodsWithMatchingAffinityTerm` looks like a suspect, which checks if PodAffinity of each unscheduled Pod matches with a newly scheduled Pod.
https://github.com/kubernetes/kubernetes/blob/efef32652af0af08a0b9c9bc547a4dce4a95f9f5/pkg/scheduler/internal/queue/scheduling_queue.go#L1235-L1252

We should skip the expensive PodAffinity matching check in `getUnschedulablePodsWithMatchingAffinityTerm` if `pInfo` is scheduling-gated.

**Comment by alculquicondor** [comment]
IIRC, AssignedPodAdd doesn't cause a revisit of all unscheduled pods (at least, without queueing hints).

The performance issue that we found happened rather with Pod terminations.

**Comment by sanposhiho** [comment]
> IIRC, AssignedPodAdd doesn't cause a revisit of all unscheduled pods

I meant, it does through getUnschedulablePodsWithMatchingAffinityTerm.

File: test/integration/scheduler_perf/config/performance-config.yaml [hunk 1]
```diff
@@ -952,3 +952,63 @@
...
+    duration: 5s
+  - opcode: churn
+    # Keep creating nodes, which result in many cluster events (Node/Add)
+    # and each of them will be processed by the scheduling queue.
+    # But, the scheduling throughput should not be impacted by the number of gated Pods.
+    mode: create
+    number: 10000
+    intervalMilliseconds: 10
+    templatePaths:
+    - config/templates/node-default.yaml
```
Comment (sanposhiho) [inline test/integration/scheduler_perf/config/performance-config.yaml:979]: If we stick to reproducing the degradation with Pod/delete, we can actually just change here to `mode: recreate` and `templatePaths: /path/to/some-pod-file.yaml`.

File: test/integration/scheduler_perf/config/performance-config.yaml [hunk 2]
```diff
@@ -952,3 +952,63 @@
...
+    countParam: $measurePods
+    collectMetrics: true
+  workloads:
+  - name: 1000Node
+    labels: [performance, fast]
+    params:
+      gatedPods: 10000
+      measurePods: 20000
+
+- name: SchedulingGatedPodsWithPodAffinityImpactForThroughput
```
Comment (sanposhiho) [inline test/integration/scheduler_perf/config/performance-config.yaml:990]: This test case needs to be added in any case as it's not overlapped with https://github.com/kubernetes/kubernetes/pull/125504.

**Review by sanposhiho — COMMENTED** [review]
@alculquicondor

**Comment by sanposhiho** [comment]
/hold

I'm convinced by @macsko (https://github.com/kubernetes/kubernetes/pull/125504#discussion_r1643309681) and will remove the first test case, which is overlapping with their PR.

**Comment by sanposhiho** [comment]
I removed the duplicate test case when rebasing (and, I should have /unhold-ed afterwards)

/unhold

**Comment by sanposhiho** [comment]
/retitle add test cases with gated pod-affinity pods to scheduler_perf

File: test/integration/scheduler_perf/config/templates/gated-pod.yaml [hunk 3]
```diff
@@ -4,7 +4,7 @@ metadata:
   generateName: gated-pod-
 spec:
   schedulingGates:
-  - name: test.k8s.io/hold
+    - name: test.k8s.io/hold
   containers:
-  - image: registry.k8s.io/pause:3.10
-    name: pause
+    - image: registry.k8s.io/pause:3.10
+      name: pause
```
[the remark this answers is not in the document]
Comment (sanposhiho) [reply test/integration/scheduler_perf/config/templates/gated-pod.yaml:10]: Okay, looks like my yaml formatter did this job.

**Comment by sanposhiho** [comment]
@alculquicondor Updated with your suggestions.

**Review by alculquicondor — COMMENTED** [review]
/lgtm
/approve

**Pull request merged**